### ISSUE TITLE
[구현] 메인어코드 추출 Machine Learning 모델 학습

### ISSUE THREAD

**Issue by ParkVitna** [skn-ai14-250409/SKN14-Final-2Team#43]
## ✔️ 상세 내용

> 다중 분류 머신러닝 모델 학습을 위한 학습데이터셋 정리 (~8/28)
> 모델 학습 및 평가지표 비교 분석 + 산출물 초안 (~8/29)
> 다중 분류 머신러닝 모델 성능평가지표 점수 향상을 위한 추가 개발 (~8/29)

## 🎯 이유/목적/배경
> 사용자 입력 자연어 향 설명에서 향수의 메인 어코드를 뽑기 위한 머신러닝 학습 데이터 확정
> 다중 분류 모델 학습
> 금주 산출물 내용 작성(인공지능 데이터 전처리 결과서, 인공지능 학습 결과서, 학습된 인공지능 모델)

## 📆 할 일

- [x] 머신러닝 다중 분류 모델 공부
- [x] ~~현재 우리가 가지고 있는 perfume_final.csv에서 사용할 수 있을 방법~~ → 데이터 너무 적음
- [x] Kaggle 등 머신러닝을 학습시킬 새로운 데이터셋 찾아보기
- [x] 다중 분류 모델 중 우리 프로젝트에 활용 가능한 임베딩 모델 찾기
- [x] 다중 분류 모델 중 우리 프로젝트에 활용 가능한 분류기 찾기
- [x] 새로운 데이터셋으로 다중 라벨 텍스트 분류 → 훈련/평가/추론
- [x] 데이터셋이 정해지면 여러 분류 모델을 각각 학습 시킨 후, 동일한 평가지표로 비교/분석
- [x] ML 학습시킨 모델 중 제일 성능 좋은 모델 기준으로 .pt, .pkl 저장

## 🎁 작업 결과

> 구현 완료된 결과(글/사진/그림 등), 장애사항도 있었다면 같이 작성
-  TF-IDF + One-vs-Rest 로지스틱 (장: 구현 간단, 속도 빠름, 적은 리소스 / 단: 한국어 입력 직접 처리 불리 번역 후 추론 방식으로 해결 필요)
- ✅멀티링구얼 임베딩 + 선형 분류기 (장: 영어로 학습해도 한국어 입력을 그대로 넣어 추론 가능 / 단 : 임베딩 계산 시간이 조금 더 소요됨)

> 데이터: 향수 설명 텍스트 1394개 (멀티라벨 분류)
> 임베딩 모델: BERT계열 ✅paraphrase-multilingual-MiniLM-L12-v2 , distiluse-base-multilingual-cased-v2 , sentence-transformers/LaBSE , xlm-roberta-base (SentenceTransformer)
> 분류기: Logistic Regression (OneVsRest), XGBoost, AutoModelForSequenceClassification(huggingface 모델, 딥러닝 기반), ✅VotingClassifier, StackingClassifier(앙상블)
> 평가: ✅Threshold 기반 + Top-K 기반
> 설정 기본값 : THRESHOLD = 0.35 , TOP_K = 3, RARE_MIN_COUNT = 10 (희소 라벨 제거)

- 평가지표(F1) 높이기 위한 방법 (진행중)
1. Threshold Grid Search (0.2 ~ 0.5 튜닝) → [Best Threshold] 0.50 → Micro-F1=0.3497
2. 분류기 XGBoost : 멀티라벨에서 좋은 성능 기대
3. 라벨마다 최적 threshold를 찾기
4. n_estimators(=epoch) 추가 
5. 라벨 정제 단계에서 "of", "the" 제거
6. 임베딩 모델 교체① : sentence-transformers/LaBSE + 라벨마다 최적 threshold를 찾기 + 라벨 정제 단계
7. 임베딩 모델 교체② : xlm-roberta-base + 라벨마다 최적 threshold를 찾기 + 라벨 정제 단계 + 직접 SentenceTransformer에 구성 가능한 형태로 Pooling까지 설정
8. 분류기 교체① : paraphrase-multilingual-MiniLM-L12-v2 + AutoModelForSequenceClassification + 라벨마다 최적 threshold를 찾기 + 라벨 정제 단계 + Epoch 8
9. ~~description 설명 전처리 추가(브랜드명 제거, 불용어 제거, 불필요한 특수문자 처리) 문맥 파악 불가로 성능 더 떨어짐~~ + paraphrase-multilingual-MiniLM-L12-v2 + XGBoost OvR / AutoModelForSequenceClassification
10. 분류기 VotingClassifier,  StackingClassifier

> 현재 모델 파이프라인은 MiniLM 임베딩(딥러닝) + VotingClassifier(머신러닝) 구조라서, 저장을 두 가지로 나눠야 해요
> 딥러닝 MiniLM 모델은 PyTorch 모델이므로 → .pt 파일로 저장 (torch.save)
> VotingClassifier + OneVsRestClassifier + MultiLabelBinarizer는 sklearn 객체이므로 → .pkl 파일로 저장 (joblib.dump)
> 노이즈 라벨 제거하여 저장 minilm_model.pt, label_info.pkl
https://drive.google.com/drive/folders/1Oz5egTL1pcBcZxFV8LXyRE927vWwlTzJ?usp=sharing

## 🔗 추가 정보/참고 자료
> 관련 문서, 스크린샷, 링크 등 첨부,
> [perfume_final.csv](https://github.com/user-attachments/files/21998996/perfume_final.csv)
> https://huggingface.co/datasets/doevent/perfume → description 향설명 , fragrances 메인어코드 [12.1 MB]

```
paraphrase-multilingual-MiniLM-L12-v2 | ✅문장 유사도 최적화된 경량 BERT (다국어)
distiluse-base-multilingual-cased-v2 | DistilBERT 기반 다국어 문장 임베딩 모델 (BERT보다 가볍고 빠름)
```

```
== F1 점수 (최소 0.6 이상, 0.7 이상이여야 그나마 쓸만한 모델) ==
Micro-F1 (0.3211 vs 0.3105) : 전체 라벨/샘플을 통틀어 TP/FP/FN을 한 번에 합산해서 계산
자주 등장하는 ‘강라벨(헤드)’의 영향이 큼 → 전반적인 운영 지표로 자주 씀

Macro-F1 (0.1823 vs 0.2028) : 라벨별 F1의 산술 평균(모든 라벨을 동일 가중)
드문 ‘꼬리 라벨(테일)’**에 취약하면 크게 떨어짐. 값이 낮다는 건 라벨 불균형 영향이 큼을 시사

Sample-F1 (0.3262 vs 0.3085) : 샘플(문장)별로 F1을 구해 평균
사용자의 체감(한 문장에서 예측이 얼마나 맞았나)에 가까운 지표
```

- 임베딩 방식 비교 표 (TF-IDF vs Word2Vec vs FastText vs BERT vs SBERT)
<img width="1076" height="548" alt="Image" src="https://github.com/user-attachments/assets/1a9d809b-cf04-49dc-b3f5-1ee51cc0025e" />

**Comment by ParkVitna** [comment]
Model : distiluse-base-multilingual-cased-v2

[Embed] train: (21001, 512) | time: 17.16s | thru: 1223.8/s
[Embed] valid: (5251, 512) | time: 4.25s | thru: 1236.3/s
[Train] OvR-LogReg: 59.70s

=== Threshold-based ===
Micro-F1: 0.3211
Macro-F1: 0.1823
Sample-F1: 0.3262

[classification_report @thr]
```
              precision    recall  f1-score   support

       Amber       0.40      0.88      0.55      1738
    Aromatic       0.19      0.86      0.31       450
     Blossom       0.02      0.44      0.04        25
     Bouquet       0.02      0.57      0.04        47
   Carnation       0.00      0.00      0.00         2
      Citrus       0.22      0.85      0.35       981
   Classical       0.31      0.84      0.45      1313
       Crisp       0.19      0.87      0.32       858
         Dry       0.09      0.83      0.16       260
      Floral       0.59      0.88      0.71      2141
      Flower       0.09      0.81      0.17       329
     Fougère       0.19      0.86      0.31       450
       Fresh       0.01      0.23      0.01        22
     Fresher       0.59      0.87      0.71      2838
      Fruity       0.24      0.86      0.37       899
    Gardenia       0.02      0.80      0.04         5
    Gourmand       0.11      0.79      0.19       326
       Green       0.08      0.76      0.15       314
        Iris       0.03      0.58      0.06        36
     Jasmine       0.09      0.56      0.15        36
        Lily       0.02      0.62      0.04         8
    Magnolia       0.06      0.75      0.11         4
      Mimosa       0.10      0.29      0.15         7
       Mossy       0.06      0.76      0.11       251
        Musk       0.03      0.55      0.06        82
      Orange       0.02      0.44      0.04        25
    Oriental       0.00      0.00      0.00         1
        Rich       0.01      0.49      0.03        41
      Richer       0.06      0.83      0.11       179
        Rose       0.13      0.82      0.22       182
        Soft       0.12      0.84      0.21       489
       Spicy       0.02      0.67      0.04        42
    Tuberose       0.03      0.46      0.06        26
      Valley       0.02      0.50      0.03         6
      Violet       0.05      0.75      0.09         4
       Water       0.07      0.80      0.14       166
       White       0.09      0.81      0.17       329
       Woods       0.28      0.88      0.43      1138
       Woody       0.19      0.86      0.31       683
          of       0.02      0.50      0.03         6
         the       0.02      0.50      0.03         6

   micro avg       0.20      0.85      0.32     16745
   macro avg       0.12      0.66      0.18     16745
weighted avg       0.33      0.85      0.45     16745
 samples avg       0.21      0.85      0.33     16745
```


=== Top-K-based ===
Micro-F1: 0.3105
Macro-F1: 0.2028
Sample-F1: 0.3085

[classification_report @topK]
```
              precision    recall  f1-score   support

       Amber       0.58      0.27      0.37      1738
    Aromatic       0.34      0.53      0.42       450
     Blossom       0.05      0.32      0.09        25
     Bouquet       0.04      0.28      0.06        47
   Carnation       0.00      0.00      0.00         2
      Citrus       0.43      0.25      0.31       981
   Classical       0.44      0.24      0.31      1313
       Crisp       0.35      0.24      0.29       858
         Dry       0.17      0.34      0.23       260
      Floral       0.76      0.50      0.60      2141
      Flower       0.14      0.21      0.17       329
     Fougère       0.32      0.57      0.41       450
       Fresh       0.01      0.14      0.02        22
     Fresher       0.72      0.19      0.30      2838
      Fruity       0.43      0.33      0.37       899
    Gardenia       0.06      0.60      0.11         5
    Gourmand       0.21      0.40      0.27       326
       Green       0.18      0.27      0.22       314
        Iris       0.08      0.50      0.14        36
     Jasmine       0.20      0.36      0.25        36
        Lily       0.08      0.50      0.14         8
    Magnolia       0.08      0.25      0.12         4
      Mimosa       0.29      0.29      0.29         7
       Mossy       0.10      0.18      0.13       251
        Musk       0.05      0.27      0.09        82
      Orange       0.06      0.32      0.10        25
    Oriental       0.00      0.00      0.00         1
        Rich       0.01      0.15      0.03        41
      Richer       0.09      0.28      0.13       179
        Rose       0.34      0.60      0.44       182
        Soft       0.25      0.21      0.23       489
       Spicy       0.04      0.43      0.07        42
    Tuberose       0.11      0.35      0.17        26
      Valley       0.00      0.00      0.00         6
      Violet       0.12      0.50      0.20         4
       Water       0.17      0.54      0.25       166
       White       0.15      0.15      0.15       329
       Woods       0.45      0.23      0.31      1138
       Woody       0.30      0.29      0.30       683
          of       0.05      0.50      0.10         6
         the       0.07      0.50      0.13         6

   micro avg       0.32      0.30      0.31     16745
   macro avg       0.20      0.32      0.20     16745
weighted avg       0.47      0.30      0.34     16745
 samples avg       0.32      0.31      0.31     16745
```

[Example Prediction]
['Water', 'Crisp', 'Fresher']

**Comment by ParkVitna** [comment]
분류기: XGBoost

=== Threshold-based ===
Micro-F1: 0.4613
Macro-F1: 0.1934
Sample-F1: 0.4435

=== Top-K-based ===
Micro-F1: 0.4825
Macro-F1: 0.1919
Sample-F1: 0.4809

**Comment by ParkVitna** [comment]
다중 라벨 분류 · XGBoost · One-vs-Rest) 전부 반영

**Comment by ParkVitna** [comment]
xlm-roberta-base 임베딩 + XGBoost OvR + Threshold 최적화(runpod_train_multilingual_fragrance_06.ipynb)

[Best Thresholds per label]
```
Amber: 0.20
Aromatic: 0.20
Blossom: 0.20
Bouquet: 0.20
Carnation: 0.20
Citrus: 0.22
Classical: 0.20
Crisp: 0.20
Dry: 0.22
Floral: 0.28
Flower: 0.24
Fougère: 0.20
Fresh: 0.20
Fresher: 0.24
Fruity: 0.20
Gardenia: 0.20
Gourmand: 0.20
Green: 0.20
Iris: 0.20
Jasmine: 0.20
Lily: 0.20
Magnolia: 0.20
Mimosa: 0.20
Mossy: 0.22
Musk: 0.20
Orange: 0.20
Oriental: 0.20
Rich: 0.20
Richer: 0.20
Rose: 0.20
Soft: 0.22
Spicy: 0.38
Tuberose: 0.20
Valley: 0.20
Violet: 0.20
Water: 0.20
White: 0.24
Woods: 0.20
Woody: 0.20
of: 0.20
the: 0.20
```

=== Threshold-based ===
Micro-F1: 0.4790
Macro-F1: 0.1518
Sample-F1: 0.4754

[classification_report @thr]
```
              precision    recall  f1-score   support

       Amber       0.40      0.83      0.54      1738
    Aromatic       0.41      0.22      0.29       450
     Blossom       0.00      0.00      0.00        25
     Bouquet       0.25      0.02      0.04        47
   Carnation       0.00      0.00      0.00         2
      Citrus       0.32      0.34      0.33       981
   Classical       0.36      0.60      0.45      1313
       Crisp       0.30      0.30      0.30       858
         Dry       0.35      0.05      0.09       260
      Floral       0.59      0.80      0.68      2141
      Flower       0.39      0.02      0.04       329
     Fougère       0.41      0.22      0.29       450
       Fresh       0.00      0.00      0.00        22
     Fresher       0.56      0.97      0.71      2838
      Fruity       0.33      0.40      0.36       899
    Gardenia       0.00      0.00      0.00         5
    Gourmand       0.37      0.10      0.15       326
       Green       0.26      0.04      0.06       314
        Iris       0.00      0.00      0.00        36
     Jasmine       1.00      0.03      0.05        36
        Lily       0.00      0.00      0.00         8
    Magnolia       1.00      0.25      0.40         4
      Mimosa       0.00      0.00      0.00         7
       Mossy       0.43      0.01      0.02       251
        Musk       0.00      0.00      0.00        82
      Orange       0.00      0.00      0.00        25
    Oriental       0.00      0.00      0.00         1
        Rich       0.00      0.00      0.00        41
      Richer       0.45      0.03      0.05       179
        Rose       0.71      0.12      0.21       182
        Soft       0.25      0.05      0.08       489
       Spicy       0.33      0.02      0.04        42
    Tuberose       1.00      0.04      0.07        26
      Valley       0.00      0.00      0.00         6
      Violet       0.00      0.00      0.00         4
       Water       0.50      0.16      0.24       166
       White       0.39      0.02      0.04       329
       Woods       0.33      0.57      0.42      1138
       Woody       0.30      0.24      0.27       683
          of       0.00      0.00      0.00         6
         the       0.00      0.00      0.00         6

   micro avg       0.44      0.53      0.48     16745
   macro avg       0.29      0.16      0.15     16745
weighted avg       0.42      0.53      0.42     16745
 samples avg       0.45      0.54      0.48     16745
```

[Example Prediction]
print(predict_multilingual("바닷가에서 느껴지는 시원하고 약간 달콤한 향이 좋아요", topk=3, thresholds=thresholds))
['Classical', 'Crisp', 'Fresher', 'Fruity', 'Woods']

**Comment by ParkVitna** [comment]
runpod_train_multilingual_fragrance_10.ipynb

```
============================================
MiniLM 임베딩 + VotingClassifier (앙상블)
+ 하드코딩된 Threshold 적용(요청값 재현)
+ 개선: LR에 class_weight="balanced", 소프트보팅 가중치 조정
============================================
```

=== Threshold-based ===
Micro-F1: 0.5006
Macro-F1: 0.2501
Sample-F1: 0.4996

**Comment by ParkVitna** [comment]
runpod_train_multilingual_fragrance_11.ipynb

```
============================================
Unsloth LLaMA-3-8B-Instruct 기반
멀티라벨 분류 + Threshold 최적화

DATA_CSV = "perfumes_huggingface.csv"
MODEL_NAME = "unsloth/llama-3-8b-Instruct"   # 일반 디코더 모델
MAX_LEN = 256
BATCH_SIZE = 8
EPOCHS = 1
LR = 2e-5
============================================
```

=== Threshold-Optimized Results ===
Micro-F1: 0.4357
Macro-F1: 0.3437
Sample-F1: 0.4397
